packages/stratimikos/adapters/belos/src/Thyra_BelosLinearOpWithSolveFactory_def.hpp

```diff
@@ -459,60 +367,21 @@ Teuchos::ValidatorXMLConverterDB::addConverter(
       "support left preconditioning.");
     Teuchos::ParameterList
       &solverTypesSL = validParamList->sublist(SolverTypes_name);
-    {
-      Belos::BlockGmresSolMgr<Scalar,MV_t,LO_t> mgr;
-      solverTypesSL.sublist(BlockGMRES_name).setParameters(
-        *mgr.getValidParameters()
-        );
-    }
-    {
-      Belos::PseudoBlockGmresSolMgr<Scalar,MV_t,LO_t> mgr;
-      solverTypesSL.sublist(PseudoBlockGMRES_name).setParameters(
-        *mgr.getValidParameters()
-        );
-    }
-    {
-      Belos::BlockCGSolMgr<Scalar,MV_t,LO_t> mgr;
-      solverTypesSL.sublist(BlockCG_name).setParameters(
-        *mgr.getValidParameters()
-        );
-    }
-    {
-      Belos::PseudoBlockCGSolMgr<Scalar,MV_t,LO_t> mgr;
-      solverTypesSL.sublist(PseudoBlockCG_name).setParameters(
-        *mgr.getValidParameters()
-        );
-    }
-    {
-      Belos::PseudoBlockStochasticCGSolMgr<Scalar,MV_t,LO_t> mgr;
-      solverTypesSL.sublist(PseudoBlockStochasticCG_name).setParameters(
-        *mgr.getValidParameters()
-        );
-    }
-    {
-      Belos::GCRODRSolMgr<Scalar,MV_t,LO_t> mgr;
-      solverTypesSL.sublist(GCRODR_name).setParameters(
-        *mgr.getValidParameters()
-        );
-    }
-    {
-      Belos::RCGSolMgr<Scalar,MV_t,LO_t> mgr;
-      solverTypesSL.sublist(RCG_name).setParameters(
-        *mgr.getValidParameters()
-        );
-    }
-    {
-      Belos::MinresSolMgr<Scalar,MV_t,LO_t> mgr;
-      solverTypesSL.sublist(MINRES_name).setParameters(
-        *mgr.getValidParameters()
-        );
-    }
-    {
-      Belos::TFQMRSolMgr<Scalar,MV_t,LO_t> mgr;
-      solverTypesSL.sublist(TFQMR_name).setParameters(
-        *mgr.getValidParameters()
-        );
+
+    validParamList->set(SolverType_name, SolverType_default);
+
+    Belos::ThyraSolverFactory<Scalar> factory;
+    Teuchos::Array<std::string> supportedSolvers = factory.supportedSolverNames();
+    typedef Belos::SolverManager<Scalar,MV_t,LO_t> IterativeSolver_t;
+    for (size_t i = 0; i < supportedSolvers.size(); i++) {
+      try {
+        RCP<IterativeSolver_t> iterativeSolver = factory.create ( supportedSolvers[i], Teuchos::null );
+        solverTypesSL.sublist(supportedSolvers[i]) = *(iterativeSolver->getValidParameters());
+      } catch (std::invalid_argument) {
+        // pass
```

Comment:
Hm, this is using try-catch as control flow.  I guess there's no way around it, though.